package-name-map/src/package-name-map.ts

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A Scope and also the top-level interface for package name mappings.
+ */
+export interface Scope {
+  path_prefix?: string;
+  packages?: {[name: string]: Package};
+  scopes?: {[path: string]: Scope};
+}
+
+export interface Package {
+  main: string;
+  path?: string;
+}
+
+/**
+ * A path/Scope pair for use in the internal algorithm.
+ */
+interface ScopeEntry {
+  path: string;
+  scope: Scope;
+}
+
+/**
+ * The result of the FindPackageEntry operation.
+ */
+interface FindPackageResult {
+  scopeContext: ScopeEntry[];
+  packageName?: string;
+  package?: Package;
+}
+
+/**
+ * An object that can resolve specifiers using its package name map
+ * configuration object.
+ *
+ * The schema of the configuration object is defined in the proposal.
+ *
+ * Currently this class can only have one mapping, but a future change could
+ * support multiple mappings, added over time, as discussed in
+ * https://github.com/domenic/package-name-maps/issues/14
+ *
+ * This class doesn't yet implement any validation of the mapping. We need to
+ * decide whether validation errors should be early or late errors. Either way,
+ * it's still a TODO to implement a validate() method.
+ */
+export class PackageNameMap {
+  private _map: Scope;
+
+  constructor(map: Scope) {
+    this._map = map;
+  }
+
+  resolve(specifier: string, referrerURL: string): string {
+    // 1. If specifier parses as absolute URL, return the specifier.
+    try {
+      new URL(specifier);
+      return specifier;
+    } catch (e) {
+      // proceed
+    }
+
+    // 2. If specifier starts with `./`, `../`,  or `/`, return the specifier.
+    if (/^\.{0,2}\//.test(specifier)) {
+      return specifier;
```

Comment:
Shouldn't this return `new URL(specifier, referrerURL).href`?

Reply:
done